fiftyone/core/map/threading.py

```diff
@@ -0,0 +1,225 @@
+"""
+Threading mapping backend
+
+| Copyright 2017-2025, Voxel51, Inc.
+| `voxel51.com <https://voxel51.com/>`_
+|
+"""
+
+import concurrent.futures
+import queue
+import threading
+from typing import (
+    Callable,
+    Iterator,
+    List,
+    Literal,
+    Optional,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+)
+
+import bson
+import fiftyone.core.config as focc
+import fiftyone.core.map.batcher as fomb
+import fiftyone.core.map.mapper as fomm
+import fiftyone.core.utils as fou
+from fiftyone.core.map.typing import SampleCollection
+from tqdm import tqdm
+
+T = TypeVar("T")
+R = TypeVar("R")
+
+ResultQueue = queue.Queue[
+    Tuple[bson.ObjectId, Union[Exception, None], Union[R, None]]
+]
+
+
+class ThreadMapper(fomm.LocalMapper):
```

Comment:
Any way we can DRY these 2 implementations? They've got a lot of similarities.

Reply:
Maybe a little bit. They use different synchronization primitives and the two methods do things slight differently. It might make it more confusing to read and work in the name of being DRY.